pkg/controllers/machine/disruption/drift.go

```diff
@@ -12,62 +12,58 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package node
+package disruption
 
 import (
 	"context"
 	"fmt"
 	"time"
 
-	"github.com/samber/lo"
-	v1 "k8s.io/api/core/v1"
 	"knative.dev/pkg/logging"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	"github.com/aws/karpenter-core/pkg/apis/settings"
 	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
 	"github.com/aws/karpenter-core/pkg/cloudprovider"
-	"github.com/aws/karpenter-core/pkg/utils/machine"
 )
 
 type Drift struct {
-	kubeClient    client.Client
 	cloudProvider cloudprovider.CloudProvider
 }
 
-func (d *Drift) Reconcile(ctx context.Context, provisioner *v1alpha5.Provisioner, node *v1.Node) (reconcile.Result, error) {
-	// If the node is marked as voluntarily disrupted by another controller, do nothing.
-	val, hasAnnotation := node.Annotations[v1alpha5.VoluntaryDisruptionAnnotationKey]
-	if hasAnnotation && val != v1alpha5.VoluntaryDisruptionDriftedAnnotationValue {
+func (d *Drift) Reconcile(ctx context.Context, _ *v1alpha5.Provisioner, machine *v1alpha5.Machine) (reconcile.Result, error) {
+	if !machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched).IsTrue() {
+		return reconcile.Result{}, nil
+	}
+	// If the machine is marked as voluntarily disrupted by another controller, do nothing.
+	voluntarilyDisrupted := machine.StatusConditions().GetCondition(v1alpha5.MachineVoluntarilyDisrupted)
+	if voluntarilyDisrupted.IsTrue() && voluntarilyDisrupted.Reason != v1alpha5.VoluntarilyDisruptedReasonDrifted {
 		return reconcile.Result{}, nil
 	}
 
 	// From here there are three scenarios to handle:
 	// 1. If drift is not enabled but the node is drifted, remove the annotation
 	//    so another disruption controller can annotate the node.
 	if !settings.FromContext(ctx).DriftEnabled {
-		if val == v1alpha5.VoluntaryDisruptionDriftedAnnotationValue {
-			delete(node.Annotations, v1alpha5.VoluntaryDisruptionAnnotationKey)
-			logging.FromContext(ctx).Debugf("removing drift annotation from node as drift has been disabled")
+		if voluntarilyDisrupted.IsTrue() {
+			_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineVoluntarilyDisrupted)
+			logging.FromContext(ctx).Debugf("removing drift status condition from machine as drift has been disabled")
```

Comment:
It seems like the condition we're removing is more general than "drift status"

Reply:
We check above that the statusCondition reason (if set) is equal to whichever disruption mechanism we are validating (in this case drift) and return early if it isn't. That way, when we get to this condition, if it is True, we know that it's because of drift so we know we are safe to clear the condition

Reply:
That makes sense, thanks. And because we're serializing all the reconciliers that may set values for this common condition, we can be confident we aren't racing with other actors mutating the condition values as the flow progresses in this one reconciliation.

Could we make this a little more readable by adding a new var after that short-circuit when it's disrupted by another mechanism like this?:

```
 	if voluntarilyDisrupted.IsTrue() && voluntarilyDisrupted.Reason != v1alpha5.VoluntarilyDisruptedReasonDrifted {
		return reconcile.Result{}, nil
	}
        // If we get this far we know that any voluntarilyDisrupted status marker is our own
        hasDriftedStatus := voluntarilyDisrupted.IsTrue()
```

I think if the rest of the flow as-is uses something like `hasDriftedStatus` rather than `voluntarilyDisrupted.IsTrue()` it's a little easier to read what's going on and why as a human.

Reply:
That's a good callout 👍